### DISCUSSION TITLE
Blaze persistence and Hibernate vector/pgvector

### DISCUSSION THREAD

**Discussion by Andreliima** [Blazebit/blaze-persistence#1879] [Q&A] [answered]
Hello,

I would like to use Blaze persistence to do similarity search using Hibernate vector (pgvector in postgres), but for a reason i couldn't figure out,  it keeps failing on doing implicit group by. Querying like this:

    float[] vectors = new float[]{...};
    new BlazeJPAQuery<>(entityManager,customJPQLNextTemplates, criteriaBuilderFactory)
                    .select(Projections.constructor(ImageSearchDto.class,
                                    image.id,
                                    image.path,
                                    image.starred,
                                    Projections.constructor(ImageUrlsDto.class,
                                            image.sourceSmall,
                                            image.sourceMedium,
                                            image.sourceLarge)))
                    .from(image)
                    .where(predicate)
                    .orderBy(new OrderSpecifier<>(Order.ASC, Expressions.template(Float.class, "l2_distance({0}, {1})", image.embedding, vectors)))
                    .fetchPage((int) pageable.getOffset(), pageable.getPageSize());

throws error:

    "Can't use the expression 'l2_distance(image.embedding,:param_1)' as an implicit group by clause, because the parameter ':param_1' with the value '[F@165362df' can't be rendered as literal which is required!"

Same query without l2_distance expression in orderBy, but included in select statement works.

Image.embedding is defined as:

    @Column(name = "embedding")
    @JdbcTypeCode(SqlTypes.VECTOR)
    @Array(length = 512)
    private float[] embedding;

Any pointers/ideas how to get this working?

**Comment by beikov** [comment]
Hi there,
I don't understand why Blaze-Persistence thinks that there is a need for grouping. I suppose a function that you are using is aggregating.

Could you please try to create a reproducer for me to look into? This should just work fine. In the meantime, you can turn off implicit group by generation for the order by clause with the [`com.blazebit.persistence.implicit_group_by_from_order_by` configuration property](https://persistence.blazebit.com/documentation/1.6/core/manual/en_US/index.html#IMPLICIT_GROUP_BY_FROM_ORDER_BY), though I'm not sure how you can apply that in a QueryDSL query. Maybe @jwgmeligmeyling knows more?

**Comment by Andreliima** [comment]
Heya,
Unfortunately that configuration property didn't do the trick.

I created a quick sample app that should exhibit that same error:
[https://github.com/Andreliima/blaze-l2distance-test](https://github.com/Andreliima/blaze-l2distance-test)

There is a single test that reproduces it.

**Reply by sinissaar** [reply]
@beikov have you had the time to look at this reproduction? We're still facing this issue...

**Reply by jwgmeligmeyling** [reply]
Could you try registering `l2_distance` as explicit non-aggregate function in the Blaze Criteria Configuration?


```java
    public static CriteriaBuilderConfiguration getConfiguration() {
        CriteriaBuilderConfiguration config = Criteria.getDefault();
        config.registerFunction(new JpqlFunctionGroup("l2_distance", false));
        return config;
    }
```

The configuration properties can be set on a query level through hints:

```java
        new BlazeJPAQuery<>()
            .setHint(ConfigurationProperties.IMPLICIT_GROUP_BY_FROM_HAVING, false)
```

It won't make a difference here as setting the property globally already didn't have an effect.

**Reply by Andreliima** [reply]
@beikov @jwgmeligmeyling 

Tried registering function as non-aggregate  and also setting IMPLICIT_GROUP_BY_FROM_HAVING and IMPLICIT_GROUP_BY_FROM_ORDER_BY as false: 
```
    public CriteriaBuilderFactory createCriteriaBuilderFactory() {
        CriteriaBuilderConfiguration config = Criteria.getDefault();
        config.registerFunction(new JpqlFunctionGroup("l2_distance", false));
        config.setProperty(IMPLICIT_GROUP_BY_FROM_HAVING, "false");
        config.setProperty(IMPLICIT_GROUP_BY_FROM_ORDER_BY, "false");
        return config.createCriteriaBuilderFactory(entityManagerFactory);
    }
```
and it had no effect.

Also i tried using blaze without QueryDSL:

```
    public PagedList<Image> find(){
        PaginatedCriteriaBuilder<Image> query = criteriaBuilderFactory.create(entityManager, Image.class)
                .from(Image.class, "img")
                .where("img.embedding").isNotNull()
                .orderByAsc("l2_distance(img.embedding, :text_embeddings)")
                .orderByAsc("img.id")
                .setParameter("text_embeddings", TextEmbeddings.textEmbeddings)
                .page(0, 20);
        return query.getResultList();
    }
```
and it behaves exactly the same.

Maybe I'm missing something really obvious?

**Reply by beikov** [reply, marked as answer]
I know it's late, but I was finally able to look into your issue and figured that this is a bug in the `PaginatedCriteriaBuilder` API. I created https://github.com/Blazebit/blaze-persistence/issues/2074 to track this. In the meantime, you could use `setFirstResult()`/`setMaxResults()` on a regular `CriteriaBuilder` if you want.
